### PR TITLE
K8SPXC-522 feature: Add support for runtimeClassName

### PR DESCRIPTION
**Pull request by ollevche**
[K8SPXC-522](https://jira.percona.com/browse/K8SPXC-522) [&#10088;?&#10089;](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

Added runtimeClassName field, rerun types code generation and extended e2e tests

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| big-data | passed |
| upgrade-haproxy | passed |
| recreate | passed |
| haproxy | passed |
| scaling | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| self-healing | failed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | failed |
| monitoring-2-0 | passed |
| self-healing-advanced | passed |
| demand-backup | passed |
| smart-update | passed |
| affinity | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| pitr | passed |
| users | passed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| scheduled-backup | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| haproxy | passed |
| storage | passed |
| upgrade-haproxy | passed |
| scaling | passed |
| big-data | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| self-healing | failed |
| security-context | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| restore-to-encrypted-cluster | passed |
| one-pod | passed |
| auto-tuning | passed |
| self-healing-advanced | passed |
| demand-backup | passed |
| proxysql-sidecar-res-limits | failed |
| operator-self-healing | passed |
| users | passed |
| upgrade-proxysql | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| smart-update | failed |
| pitr | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| demand-backup-encrypted-with-tls | passed |
| upgrade-consistency | passed |
| scheduled-backup | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| haproxy | passed |
| self-healing | failed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| upgrade-haproxy | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| scaling | passed |
| recreate | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| demand-backup | passed |
| pitr | passed |
| demand-backup-encrypted-with-tls | passed |
| big-data | passed |
| scheduled-backup | passed |
| upgrade-proxysql | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| smart-update | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| upgrade-consistency | passed |

File: pkg/apis/pxc/v1/pxc_types.go [hunk 1]
```diff
@@ -69,6 +69,7 @@ type PXCScheduledBackup struct {
 	ServiceAccountName string                        `json:"serviceAccountName,omitempty"`
 	Annotations        map[string]string             `json:"annotations,omitempty"`
 	PITR               PITRSpec                      `json:"pitr,omitempty"`
+	RuntimeClassName   *string                       `json:"runtimeClassName,omitempty"`
```
Comment (delgod) [inline pkg/apis/pxc/v1/pxc_types.go:72]: NodeSelector and RuntimeClassName are very connected things.
so, both items should be set on the same storage level.
it is needed to modify `BackupStorageSpec` (not `PXCScheduledBackup`).

**Review by delgod — CHANGES_REQUESTED** [review]
Please also rebase the `K8SPXC-522` branch on the main branch.
reason: self-healing test should become fixed

File: pkg/apis/pxc/v1/pxc_types.go (hunk 1 above)
Comment (ollevche) [reply pkg/apis/pxc/v1/pxc_types.go:72]: It's already rebased onto the main branch. My previous commit is the latest one from the main branch (f512d21ea3b709b8e81bd53ee98d2464f70e0a8d). Did I miss something?

Comment (ollevche) [reply pkg/apis/pxc/v1/pxc_types.go]: Moved RuntimeClassName to BackupStorageSpec

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| big-data | passed |
| storage | passed |
| haproxy | failed |
| upgrade-haproxy | passed |
| recreate | passed |
| init-deploy | passed |
| self-healing | failed |
| scaling-proxysql | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| upgrade-proxysql | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| demand-backup | passed |
| smart-update | passed |
| one-pod | passed |
| self-healing-advanced | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| operator-self-healing | passed |
| pitr | passed |
| users | passed |
| tls-issue-self | passed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | failed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

**Review by delgod — APPROVED** [review]

**Comment by qJkee** [comment]
Hi @ollevche 
Is this code works well on low k8s versions?
By saying well i mean all works well, even without implemented functionality to k8s
Like minishift 3.11/openshift 3.11?

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| scaling | passed |
| haproxy | failed |
| storage | passed |
| recreate | passed |
| big-data | passed |
| init-deploy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| limits | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| monitoring-2-0 | passed |
| self-healing-advanced | passed |
| affinity | passed |
| smart-update | passed |
| demand-backup | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| pitr | passed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| scheduled-backup | failed |

**Comment by ollevche** [comment]
> Hi @ollevche
> Is this code works well on low k8s versions?
> By saying well i mean all works well, even without implemented functionality to k8s
> Like minishift 3.11/openshift 3.11?

Hi, @qJkee

minishift 3.11/openshift 3.11 ignores runtimeClassName field so it's safe to use

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| scaling | passed |
| scaling-proxysql | passed |
| security-context | passed |
| haproxy | failed |
| big-data | passed |
| upgrade-haproxy | passed |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| pitr | passed |
| demand-backup-encrypted-with-tls | passed |
| scheduled-backup | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |

File: pkg/pxc/backup/scheduled.go [hunk 2]
```diff
@@ -24,7 +24,7 @@ func (bcp *Backup) Scheduled(spec *api.PXCScheduledBackupSchedule, strg *api.Bac
 	labels["cluster"] = bcp.cluster
 	labels["schedule"] = genScheduleLabel(spec.Schedule)
 
-	cjts, err := bcp.scheduledJob(spec, strg, labels, operatorPod.Spec.ServiceAccountName)
+	cjts, err := bcp.scheduledJob(spec, strg, labels, operatorPod.Spec.ServiceAccountName, operatorPod.Spec.RuntimeClassName)
```
Comment (delgod) [inline pkg/pxc/backup/scheduled.go:27]: NodeSelector and RuntimeClassName are very connected things.
so, both items should be set on the same storage level (not operatorPod)
```suggestion
	cjts, err := bcp.scheduledJob(spec, strg, labels, operatorPod.Spec.ServiceAccountName)
```

**Review by delgod — CHANGES_REQUESTED** [review]

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| scaling | passed |
| storage | passed |
| recreate | passed |
| haproxy | passed |
| upgrade-haproxy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| security-context | passed |
| limits | passed |
| monitoring-2-0 | passed |
| smart-update | passed |
| self-healing-advanced | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| operator-self-healing | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| demand-backup | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| pitr | passed |
| validation-hook | passed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| storage | passed |
| recreate | passed |
| big-data | passed |
| upgrade-haproxy | passed |
| scaling | passed |
| init-deploy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| limits | passed |
| upgrade-proxysql | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| affinity | passed |
| smart-update | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| pitr | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| haproxy | passed |
| scaling | passed |
| storage | passed |
| upgrade-haproxy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| security-context | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| self-healing-advanced | passed |
| tls-issue-cert-manager | passed |
| operator-self-healing | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

**Review by hors — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| self-healing-advanced | passed |
| big-data | passed |
| upgrade-haproxy | passed |
| operator-self-healing | passed |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| pitr | passed |
| haproxy | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| scaling | passed |
| scaling-proxysql | passed |
| security-context | passed |
| demand-backup-encrypted-with-tls | passed |
| scheduled-backup | passed |

**Comment by JNKPercona** [comment]
PXC operator docker - perconalab/percona-xtradb-cluster-operator:PR-746-5fcd7bba

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| haproxy | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| upgrade-haproxy | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| scaling | passed |
| tls-issue-self | passed |
| scaling-proxysql | passed |
| tls-issue-cert-manager | passed |
| recreate | passed |
| security-context | passed |
| tls-issue-cert-manager-ref | passed |
| restore-to-encrypted-cluster | passed |
| validation-hook | passed |
| demand-backup | passed |
| pitr | passed |
| demand-backup-encrypted-with-tls | passed |
| scheduled-backup | passed |
| big-data | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |

**Review by hors — APPROVED** [review]

**Review by qJkee — APPROVED** [review]

**Pull request merged**